File: app/src/main/java/ua/hospes/nfs/marathon/data/team/TeamsRepositoryImpl.java
package ua.hospes.nfs.marathon.data.team;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;
import ua.hospes.nfs.marathon.data.drivers.DriversRepositoryImpl;
import ua.hospes.nfs.marathon.data.team.mapper.TeamsMapper;
import ua.hospes.nfs.marathon.data.team.storage.TeamsDbStorage;
import ua.hospes.nfs.marathon.domain.drivers.DriversRepository;
import ua.hospes.nfs.marathon.domain.drivers.models.Driver;
import ua.hospes.nfs.marathon.domain.team.TeamsRepository;
import ua.hospes.nfs.marathon.domain.team.models.Team;

/**
 * @author Andrew Khloponin
 */
@Singleton
public class TeamsRepositoryImpl implements TeamsRepository {
    private final TeamsDbStorage dbStorage;
    private final DriversRepository driversRepository;


    @Inject
    public TeamsRepositoryImpl(TeamsDbStorage dbStorage, DriversRepositoryImpl driversRepository) {
        this.dbStorage = dbStorage;
        this.driversRepository = driversRepository;
    }


    @Override
    public Observable<Team> get() {
        return dbStorage.get()
                .flatMap(teamDb -> Observable.zip(Observable.just(teamDb), getDriverByTeamId(teamDb.getId()).toList(), TeamsMapper::map));
    }

    @Override
    public Observable<Team> get(int id) {
        return dbStorage.get(id)
                .flatMap(teamDb -> Observable.zip(Observable.just(teamDb), getDriverByTeamId(teamDb.getId()).toList(), TeamsMapper::map));
    }

    @Override
    public Observable<Team> getNotInRace() {
        return dbStorage.getNotInRace()
                .flatMap(teamDb -> Observable.zip(Observable.just(teamDb), getDriverByTeamId(teamDb.getId()).toList(), TeamsMapper::map));
    }

    @Override
    public Observable<List<Team>> listen() {
        return dbStorage.listen()
                .flatMap(teamDbs -> Observable.from(teamDbs)
                        .flatMap(teamDb -> Observable.zip(
                                Observable.just(teamDb),
                                getDriverByTeamId(teamDb.getId()).toList(),
                                TeamsMapper::map)).toList()
                );
    }

    @Override
    public Observable<Boolean> save(Team team) {
        if (team.getId() == -1) {
            return dbStorage.add(TeamsMapper.map(team)).map(result -> result.getResult() != 0);
        } else {
            return dbStorage.update(TeamsMapper.map(team)).map(result -> result.getResult() != 0);
        }
    }

    @Override
    public Observable<Boolean> delete(Team team) {
        return dbStorage.remove(TeamsMapper.map(team)).map(count -> count != 0);
    }

    @Override
    public Observable<Void> clear() {
        return dbStorage.clear();
    }


    private Observable<Driver> getDriverByTeamId(int teamId) {
        return driversRepository.getByTeamId(teamId);
    }
}

File: app/src/main/java/ua/hospes/rtm/data/TeamsRepositoryImpl.kt
package ua.hospes.rtm.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import ua.hospes.rtm.db.drivers.DriverDAO
import ua.hospes.rtm.db.team.TeamDAO
import ua.hospes.rtm.domain.team.Team
import ua.hospes.rtm.domain.team.TeamsRepository
import ua.hospes.rtm.domain.team.toDbEntity

internal class TeamsRepositoryImpl(private val dao: TeamDAO, private val driverDAO: DriverDAO) : TeamsRepository {
    override suspend fun get(): List<Team> =
            withContext(Dispatchers.IO) { dao.get().map { it.toDomain(driverDAO) } }

    override suspend fun getNotInRace(): List<Team> =
            withContext(Dispatchers.IO) { dao.getNotInRace().map { it.toDomain(driverDAO) } }


    override fun listen(): Flow<List<Team>> = dao.observe().map { list -> list.map { it.toDomain(driverDAO) } }


    override suspend fun save(team: Team) =
            withContext(Dispatchers.IO) { dao.save(team.toDbEntity(), team.drivers.map { it.id }.toLongArray()) }

    override suspend fun delete(id: Long) =
            withContext(Dispatchers.IO) { dao.delete(id) }

    override suspend fun clear() =
            withContext(Dispatchers.IO) { dao.clear() }
}